spatial/r2/box.go

```diff
@@ -0,0 +1,101 @@
+// Copyright ©2022 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package r2
+
+import "math"
+
+// Box is a 2D bounding box. Well formed
+// Boxes have Min components smaller than
+// Max components.
+type Box struct {
+	Min, Max Vec
+}
+
+// NewBox is shorthand for Box{Min:Vec{x0,y0}, Max:Vec{x1,y1}}.
+// The sides are swapped so that the resulting Box is well formed.
+func NewBox(x0, y0, x1, y1 float64) Box {
+	return Box{
+		Min: Vec{X: math.Min(x0, x1), Y: math.Min(y0, y1)},
+		Max: Vec{X: math.Max(x0, x1), Y: math.Max(y0, y1)},
+	}
+}
+
+// Size returns the size of the Box.
+func (a Box) Size() Vec {
+	return Sub(a.Max, a.Min)
+}
+
+// Center returns the center of the Box.
+func (a Box) Center() Vec {
+	return Add(a.Min, Scale(0.5, a.Size()))
+}
+
+// IsEmpty returns true if a Box's volume is zero
+// or if a Min component is greater than its Max component.
+func (a Box) Empty() bool {
+	return a.Min.X >= a.Max.X || a.Min.Y >= a.Max.Y
+}
+
+// Vertices returns a slice of the 4 vertices
+// corresponding to each of the Box's corners.
+//
+// The order of the vertices is CCW in the XY plane
+// starting at the box minimum.
+func (a Box) Vertices() []Vec {
+	return []Vec{
+		0: a.Min,
+		1: {a.Max.X, a.Min.Y},
+		2: a.Max,
+		3: {a.Min.X, a.Max.Y},
+	}
+}
+
+// Union returns a box enclosing both the receiver and argument Boxes.
+func (a Box) Union(b Box) Box {
+	if a.Empty() {
+		return b
+	}
+	if b.Empty() {
+		return a
+	}
+	return Box{
+		Min: minElem(a.Min, b.Min),
+		Max: maxElem(a.Max, b.Max),
```

Comment:
If the `Box` is incorrectly constructed, this will fail to give the right answer. See `image.Rectangle`'s implementation. I would say that a box has an invariant that min <= max for both dimensions, but this is not currently guaranteed.

Reply:
My understanding is this does the same thing as image.Union. both implementations discard malformed boxes with `Empty` and then return the element wise minimum and maximum.

Reply:
Yeah, OK. With the addition of a cononicalisation method this would be OK.